### PR TITLE
test: move sync_blocks and sync_mempool functions to test_framework.py

### LINKED ISSUE #18930: test: Remove sync_blocks global
sync_blocks has been made a member of the test framework in https://github.com/bitcoin/bitcoin/pull/15773, but the global helper has been left intact for now.

It has been suggested to move the implementation to the member function as well https://github.com/bitcoin/bitcoin/pull/15773#pullrequestreview-224235071

Implementing the functionality in a single place in the test framework allows the method to take account of contextual information, such as global timeout modifications. Also, having a global that does almost-but-not-quite the same thing is confusing for test writers and reviewers.

So the implementation should be moved, along with removing the global helper.

#### Useful skills:

Python3, basic understanding of the test framework

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by ycshao**
This PR moves `sync_blocks` and `sync_mempool` out from `test_framework/util.py` to `test_framework/test_framework.py` so they can take contextual information of test framework into account.

* Change all reference callers to call functions from `test_framework.py`
* Remove `**kwargs` which is not used
* Take into account of `timeout_factor` when respecting timeout in function implementations. 
* Pass all tests by running `./test/functional/test_runner.py`

fixes #18930

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Nice! Concept ACK

**Comment by ycshao** [comment]
Seems like the style of those files are auto-formatted by `autopep8`. I'll leave the reviewers to comment about the format. 

One question about the implementation. Does it make sense  to move `assert (all([len(x.getpeerinfo()) for x in rpc_connections]))` up to immediately after the while loop starts? It will assert sooner and save some time and operations.

**Comment by ycshao** [comment]
Looks like `wallet_avoidreuse.py` failed in travis but it works fine on my machine. Is there a way to replay the travis failed tests?

**Comment by ycshao** [comment]
> If you consider this pull request important, please also help to review the conflicting pull requests. Ideally, start with the one that should be merged first.

Could someone explain what does conflicts mean here? Is there a logical order of PRs mentioned + mine to review and merge?

**Comment by adamjonas** [comment]
@ycshao DrahtBot crawls over open PRs and alerts you when there is a conflict with the main branch that requires a rebase or, in this case, other PRs that may lead to future conflicts indicating that you are working on similar code. DrahtBot is not sentient enough to decide the merge order. It's just raising it to your attention.

**Comment by adamjonas** [comment]
Re the format - autopep8 is making the diff much larger than it needs to be. For the sake of the reviewers, I'd suggest you keep the diff as minimal as possible so that they can focus on the changes relevant to your patch.

**Comment by ycshao** [comment]
@adamjonas thanks for the answer and suggestion. I'll squash and break formatting into it's own commit.

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -542,15 +564,53 @@ def join_network(self):
...
-    def sync_mempools(self, nodes=None, **kwargs):
-        sync_mempools(nodes or self.nodes, **kwargs)
+    def sync_blocks(self, nodes=None, wait=1, timeout=60):
+        """
+        Wait until everybody has the same tip.
 
-    def sync_all(self, nodes=None, **kwargs):
-        self.sync_blocks(nodes, **kwargs)
-        self.sync_mempools(nodes, **kwargs)
+        sync_blocks needs to be called with an rpc_connections set that has least
```
Comment (gzhao408) [inline test/functional/test_framework/test_framework.py:571]: I have a question - it's probably not a big deal to require at least one node to be updated, but would it be appropriate to have an optional `blockhash` parameter and enforce that hash instead?

**Review by glozow — COMMENTED** [review]
Oo first PR? Concept ACK :) I'm guessing "contextual information of test framework" is the `timeout_factor`? 
+1 to Jonas, just appeasing test/lint/lint-python.sh for now would be better.
> Does it make sense to move `assert (all([len(x.getpeerinfo()) for x in rpc_connections]))` up to immediately after the while loop starts? It will assert sooner and save some time and operations.

I had this thought too. But I think it would fail if the nodes happened to all be synced up (despite not being connected right now) before the first iteration of the while loop. I think it's possible...?

**Comment by ycshao** [comment]
> Oo first PR? Concept ACK :) I'm guessing "contextual information of test framework" is the `timeout_factor`?

Yeah for now `timeout_factor` is easiest to be used. There are other things potentially can be used. 
* `rpc_timeout` - I think `rpc_timeout` is usually sometime different than timeout people want to configure for syncing operations. Unless there is a factor that can be applied to `rpc_timeout` to estimate syncing timeout. Open to suggestions. 
* `nodes` - We can by default sync all nodes and take in a parameter in `sync_blocks` and `sync_mempool` in case subset of nodes need to be synced. This way we get rid of `nodes` input of those two functions. 

> +1 to Jonas, just appeasing test/lint/lint-python.sh for now would be better.
 
`lint-python.sh` gives me other linting errors which is not related to the file I touched. I guess I'll just make those five files I touched to be pep8 in a separate commit. If it's not necessary, it's easy to take it out.
```
test/functional/p2p_compactblocks.py:688:43: E741 ambiguous variable name 'l'
test/functional/p2p_compactblocks.py:694:13: E741 ambiguous variable name 'l'
test/functional/p2p_compactblocks.py:697:17: E741 ambiguous variable name 'l'
test/functional/p2p_dos_header_tree.py:37:38: E741 ambiguous variable name 'l'
test/functional/p2p_dos_header_tree.py:42:31: E741 ambiguous variable name 'l'
test/functional/p2p_dos_header_tree.py:43:55: E741 ambiguous variable name 'l'
test/functional/test_framework/messages.py:70:22: E741 ambiguous variable name 'l'
test/functional/test_framework/messages.py:142:16: E741 ambiguous variable name 'l'
test/functional/test_framework/messages.py:161:24: E741 ambiguous variable name 'l'
test/functional/test_framework/messages.py:177:23: E741 ambiguous variable name 'l'
test/functional/wallet_labels.py:149:13: E741 ambiguous variable name 'l'
test/functional/wallet_labels.py:155:13: E741 ambiguous variable name 'l'
test/lint/lint-python.sh: line 105: mypy: command not found
```

> > Does it make sense to move `assert (all([len(x.getpeerinfo()) for x in rpc_connections]))` up to immediately after the while loop starts? It will assert sooner and save some time and operations. 
>
> I had this thought too. But I think it would fail if the nodes happened to all be synced up (despite not being connected right now) before the first iteration of the while loop. I think it's possible...?

I guess it's theoretically possible.

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (ycshao) [reply test/functional/test_framework/test_framework.py:569]: I don't see a use case like that in current code. But I think one benefit your idea will give is to shorten test run time if the block we need to sync to is not usually the tip.

File: test/functional/test_framework/test_framework.py [hunk 2]
```diff
@@ -204,11 +213,13 @@ def setup(self):
...
         self.options.bitcoind = os.getenv("BITCOIND", default=fname_bitcoind)
-        self.options.bitcoincli = os.getenv("BITCOINCLI", default=fname_bitcoincli)
+        self.options.bitcoincli = os.getenv(
+            "BITCOINCLI", default=fname_bitcoincli)
 
         os.environ['PATH'] = os.pathsep.join([
             os.path.join(config['environment']['BUILDDIR'], 'src'),
-            os.path.join(config['environment']['BUILDDIR'], 'src', 'qt'), os.environ['PATH']
+            os.path.join(config['environment']['BUILDDIR'],
+                         'src', 'qt'), os.environ['PATH']
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:222]: I am not sure it this is actually more readable (we've generally been using more a limit of 120 instead of 80).
And taking into account the large number of conflicts due to the pep-8 change, maybe leave it out for now?

Comment (ycshao) [reply test/functional/test_framework/test_framework.py]: Reverted

File: test/functional/test_framework/test_framework.py [hunk 3]
```diff
@@ -541,15 +539,52 @@ def join_network(self):
...
 
-    def sync_blocks(self, nodes=None, **kwargs):
-        sync_blocks(nodes or self.nodes, **kwargs)
-
-    def sync_mempools(self, nodes=None, **kwargs):
-        sync_mempools(nodes or self.nodes, **kwargs)
-
-    def sync_all(self, nodes=None, **kwargs):
-        self.sync_blocks(nodes, **kwargs)
-        self.sync_mempools(nodes, **kwargs)
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:552]: why is this changed?

File: test/functional/test_framework/test_framework.py [hunk 4]
```diff
@@ -541,15 +539,52 @@ def join_network(self):
...
+        stop_time = time.time() + int(timeout * self.options.timeout_factor)
+        while time.time() <= stop_time:
+            best_hash = [x.getbestblockhash() for x in rpc_connections]
+            if best_hash.count(best_hash[0]) == len(rpc_connections):
+                return
+            # Check that each peer has at least one connection
+            assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
+            time.sleep(wait)
+        raise AssertionError("Block sync timed out after {}s:{}".format(
+            int(timeout * self.options.timeout_factor),
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:559]: You don't need to change this if you just set `timeout = int(timeout * self.options.timeout_factor)` at the beginning of the function

File: test/functional/test_framework/test_framework.py [hunk 5]
```diff
@@ -541,15 +539,52 @@ def join_network(self):
...
+            if pool.count(pool[0]) == len(rpc_connections):
+                if flush_scheduler:
+                    for r in rpc_connections:
+                        r.syncwithvalidationinterfacequeue()
+                return
+            # Check that each peer has at least one connection
+            assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
+            time.sleep(wait)
+        raise AssertionError("Mempool sync timed out after {}s:{}".format(
+            int(timeout * self.options.timeout_factor),
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py:581]: same

**Review by maflcko — COMMENTED** [review]
Concept ACK, but please change the commit title to

```
test: move sync_blocks and sync_mempool functions to test_framework.py
```

<!--

b593d82f7758d24fe397eebff73141f917d03ab9

File: test/functional/test_framework/test_framework.py (hunk 3 above)
Comment (ycshao) [reply test/functional/test_framework/test_framework.py:552]: I removed `**kwargs` which are not used.

Comment (MarcoFalke) [reply test/functional/test_framework/test_framework.py:552]: I think they might come in handy in the future. Is there a reason they will be unused forever?

Comment (ycshao) [reply test/functional/test_framework/test_framework.py:552]: I feel it's better to add it when it will be used in instead of putting it there up front just because some time in the future it might be used. What do you think?

File: test/functional/test_framework/test_framework.py (hunk 4 above)
Comment (ycshao) [reply test/functional/test_framework/test_framework.py]: Changed

File: test/functional/test_framework/test_framework.py (hunk 5 above)
Comment (ycshao) [reply test/functional/test_framework/test_framework.py]: changed

**Comment by MarcoFalke** [comment]
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits and change the commit title to something more meaningful: https://github.com/bitcoin/bitcoin/pull/19208#pullrequestreview-433226219

**Comment by MarcoFalke** [comment]
ACK cc84460c164bcb2a874d4f08b3a2624e5ee9ff0a , reviewed with --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space 💫

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK cc84460c164bcb2a874d4f08b3a2624e5ee9ff0a , reviewed with --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space 💫
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh0+gv+MXi8uR6BQOeLwvt+luY9+MDS6qKEVkd1x6tUrhNL4R/1/HzIbCPA0n0O
FH/bMuzMk19Fn4qkH4vGEJHPQTeT+CNEf9omuWRSrNU8pqicaBpAN0cpWcHbJSxD
A8TC5uecJ1vWLmYruW9mnSTyYFTL7maKas7yKv4JsmsLIRhMPhWnJ2VU7MrCq9Uv
UyR5VTYjiqmw3Gz24RJ9QTtJWTDJjlkNXs01ZrO1ijVX5Gak9QLx47wz60lh8wtT
XtRQKOxsBGbJfrFexpU3IdThhNTuLTx1zcULgE2S6IOBF1Lph/CrSuZLv69H3bhW
S45h7iliCQR5njyC93TB1GB56BDFGcvbFqkjP0jgB0sVdh+pRP4lcgTNRGPD8O+t
p4TbuTKMo6P1JbqB1TAzZCmWUjVLmTjReSvGlMKegA3FHTm2XyXQEuOT7KHNk2Bm
2BHyvABW/hlIB/79YlUQ59pTC8CIvjEl0JQxY5VXOARyESD9PAteaJ5aFplnHtYl
eIUZJOgR
=adom
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bea4e035a3dba2f2997391d4f35010649fa6ca21167797054925a5e15c237f47  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bea4e035a3dba2f2997391d4f35010649fa6ca21167797054925a5e15c237f47f01008f9e05f01a93fe95ef5b6ab10ac836a08fff010594ce5ad6ad56c80ec435d8aef4606f108f1045eebd031f008ed7222c780288ad00083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0106b1050e28df84d40615e3351f28c4a2108f1045eebd031f008eebccb7c190a49060083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010dbcc5612f65d79185f9f21ccef31505708f1045eebd031f0084169885ed2ca354b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010de679ec0889641286b599536c843f24a08f1045eebd031f00812810bab20f7d49a0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by ycshao** [comment]
@MarcoFalke @gzhao408 @adamjonas  thanks for reviewing and merge!
